From b21976225f55be31bde21b06a6511e12092376c1 Mon Sep 17 00:00:00 2001
From: Harshitha Prem <quic_hprem@quicinc.com>
Date: Tue, 23 Aug 2022 14:15:08 +0530
Subject: [PATCH] hostapd: Add frequency info in start ap

In case of single hw multiple band scenario, when we
bring up multiple virtual interface in various bands
simultaneously then the driver maps a random channel
to the virtual interface when start ap command is sent
as freq info is not present.

Hence, add the frequency information in start ap netlink
command so that the driver maps the appropriate channel
context by parsing it.

Signed-off-by: Harshitha Prem <quic_hprem@quicinc.com>

---
 src/drivers/driver_nl80211.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -171,7 +171,8 @@ static int nl80211_send_frame_cmd(struct
 				  const u16 *csa_offs, size_t csa_offs_len);
 static int wpa_driver_nl80211_probe_req_report(struct i802_bss *bss,
 					       int report);
-
+static int nl80211_put_freq_params(struct nl_msg *msg,
+                                   const struct hostapd_freq_params *freq);
 #define IFIDX_ANY -1
 
 static void add_ifidx(struct wpa_driver_nl80211_data *drv, int ifidx,
@@ -4885,6 +4886,9 @@ static int wpa_driver_nl80211_set_ap(voi
 		nla_nest_end(msg, spr);
 	}
 
+	if (params->freq && nl80211_put_freq_params(msg, params->freq) < 0)
+		goto fail;
+
 	if (params->freq && params->freq->he_enabled) {
 		struct nlattr *bss_color;
 
